src/main/scala/chisel3/util/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,191 @@
+// See LICENSE for license details.
+
+package chisel3.util
+
+import chisel3.MemBase
+import chisel3.core.{ChiselAnnotation, RunFirrtlTransform, annotate}
+import chisel3.internal.{Builder, InstanceId}
+import firrtl.annotations.{MemoryLoadFileType, _}
+import firrtl.ir.{Module => _, _}
+import firrtl.passes.Pass
+import firrtl.passes.memlib.DefAnnotatedMemory
+import firrtl.transforms.BlackBoxInlineAnno
+import firrtl.{CircuitForm, CircuitState, EmitCircuitAnnotation, LowForm, Transform, VerilogEmitter, WDefInstance}
+
+import scala.collection.mutable
+
+/**
+  * chisel implementation for load memory
+  * @param target        memory to load
+  * @param fileName      name of input file
+  * @param hexOrBinary   use $readmemh or $readmemb
+  */
+case class ChiselLoadMemoryAnnotation(
+  target:      InstanceId,
+  fileName:    String,
+  hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
+)
+  extends ChiselAnnotation with RunFirrtlTransform {
+
+  if(fileName.isEmpty) {
+    Builder.warning(
+      s"""LoadMemory from file annotations file empty file name"""
+    )
+  }
+
+  def transformClass: Class[LoadMemoryTransform] = classOf[LoadMemoryTransform]
+
+  def toFirrtl: LoadMemoryAnnotation = {
+    LoadMemoryAnnotation(target.toNamed.asInstanceOf[ComponentName], fileName, hexOrBinary)
+  }
+}
+
+
+object loadMemoryFromFile {
+  def apply(
+    memory: MemBase[_],
+    fileName: String,
+    hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
+  ): Unit = {
+    annotate(ChiselLoadMemoryAnnotation(memory, fileName))
+  }
+}
+
+/**
+  * This pass creates BlackBoxInlineAnno from the LoadMemoryAnnotations
+  * it does this even if the backend is not verilog.
+  *
+  * @param circuitState the target circuit state
+  */
+//TODO: (chick) better integration with chisel end firrtl error systems
+//scalastyle:off method.length cyclomatic.complexity regex
+class CreateBindableMemoryLoaders(circuitState: CircuitState) extends Pass {
+  var memoryCounter: Int = -1
```

Comment:
I'd prefer to not solve this problem right now. This transform should only run once so collisions should not be a problem. I think the name space problem for bind-able modules should be solved separately with more uses cases brought to bear. For the moment this variable is enough to prevent self collisions

Reply:
Pass eliminated, LoadMemoryTransform does all the work itself now.